File: backend/e2e/teams_test.go
package e2e

import (
	"context"
	"testing"

	"github.com/bcc-media/wayfarer/e2e/testutil"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestTeams(t *testing.T) {
	ctx := context.Background()
	dbMgr, _ := GetTestEnv()

	// Clean and seed with deterministic data
	require.NoError(t, dbMgr.Clean(ctx))
	data, err := dbMgr.Seed(ctx, 42, testutil.DefaultSeedConfig())
	require.NoError(t, err)

	// Setup user IDs - use different users for different roles
	userID := data.UserIDs[0]
	adminUserID := data.UserIDs[1]

	// Assign database roles
	require.NoError(t, dbMgr.AssignRole(ctx, adminUserID, testutil.RoleAdmin))

	// Setup test server
	router, cleanup, err := testutil.SetupTestServer(ctx, dbMgr)
	require.NoError(t, err)
	defer cleanup()

	client := testutil.NewGraphQLClient(router)
	defer client.Close()

	userToken, err := testutil.GenerateUserToken(userID)
	require.NoError(t, err)

	adminToken, err := testutil.GenerateAdminToken(adminUserID)
	require.NoError(t, err)

	// Get first project's team
	projectID := data.ProjectIDs[0]
	teamIDs := data.TeamIDs[projectID]
	require.NotEmpty(t, teamIDs, "should have seeded teams")
	teamID := teamIDs[0]

	t.Run("admin can query team by id with members", func(t *testing.T) {
		// Use admin token since members access is restricted
		resp := client.WithAuth(adminToken).MustExecute(t, `
			query GetTeam($id: ID!) {
				team(id: $id) {
					id
					name
					description
					parentProject { id name }
					members {
						user { id name }
						joinedAt
					}
				}
			}
		`, map[string]any{"id": teamID})

		require.False(t, resp.HasErrors(), "unexpected error: %s", resp.ErrorMessage())

		var result struct {
			Team struct {
				ID            string `json:"id"`
				Name          string `json:"name"`
				Description   string `json:"description"`
				ParentProject struct {
					ID   string `json:"id"`
					Name string `json:"name"`
				} `json:"parentProject"`
				Members []struct {
					User struct {
						ID   string `json:"id"`
						Name string `json:"name"`
					} `json:"user"`
					JoinedAt string `json:"joinedAt"`
				} `json:"members"`
			} `json:"team"`
		}
		require.NoError(t, resp.UnmarshalData(&result))

		assert.Equal(t, teamID, result.Team.ID)
		assert.NotEmpty(t, result.Team.Name)
		assert.Equal(t, projectID, result.Team.ParentProject.ID)
	})

	t.Run("user can query team basic info", func(t *testing.T) {
		// Users can query basic team info without members
		resp := client.WithAuth(userToken).MustExecute(t, `
			query GetTeam($id: ID!) {
				team(id: $id) {
					id
					name
					description
					parentProject { id }
				}
			}
		`, map[string]any{"id": teamID})

		require.False(t, resp.HasErrors(), "unexpected error: %s", resp.ErrorMessage())

		var result struct {
			Team struct {
				ID            string `json:"id"`
				Name          string `json:"name"`
				Description   string `json:"description"`
				ParentProject struct {
					ID string `json:"id"`
				} `json:"parentProject"`
			} `json:"team"`
		}
		require.NoError(t, resp.UnmarshalData(&result))

		assert.Equal(t, teamID, result.Team.ID)
		assert.NotEmpty(t, result.Team.Name)
	})

	t.Run("admin can list all teams via filter", func(t *testing.T) {
		resp := client.WithAuth(adminToken).MustExecute(t, `
			query GetTeams($projectId: ID!) {
				teams(filter: { projectId: $projectId }, first: 100) {
					edges {
						node {
							id
							name
							joinCode
						}
					}
					totalCount
				}
			}
		`, map[string]any{"projectId": projectID})

		require.False(t, resp.HasErrors(), "unexpected error: %s", resp.ErrorMessage())

		var result struct {
			Teams struct {
				Edges []struct {
					Node struct {
						ID       string `json:"id"`
						Name     string `json:"name"`
						JoinCode string `json:"joinCode"`
					} `json:"node"`
				} `json:"edges"`
				TotalCount int `json:"totalCount"`
			} `json:"teams"`
		}
		require.NoError(t, resp.UnmarshalData(&result))

		assert.Greater(t, result.Teams.TotalCount, 0)
	})

	t.Run("admin can create team", func(t *testing.T) {
		resp := client.WithAuth(adminToken).MustExecute(t, `
			mutation CreateTeam($projectId: ID!, $input: CreateTeamInput!) {
				createTeam(projectId: $projectId, input: $input) {
					id
					name
					description
					joinCode
					parentProject { id }
				}
			}
		`, map[string]any{
			"projectId": projectID,
			"input": map[string]any{
				"name":        "E2E Test Team",
				"description": "Created by E2E test",
			},
		})

		require.False(t, resp.HasErrors(), "unexpected error: %s", resp.ErrorMessage())

		var result struct {
			CreateTeam struct {
				ID            string `json:"id"`
				Name          string `json:"name"`
				Description   string `json:"description"`
				JoinCode      string `json:"joinCode"`
				ParentProject struct {
					ID string `json:"id"`
				} `json:"parentProject"`
			} `json:"createTeam"`
		}
		require.NoError(t, resp.UnmarshalData(&result))

		assert.NotEmpty(t, result.CreateTeam.ID)
		assert.Equal(t, "E2E Test Team", result.CreateTeam.Name)
		assert.NotEmpty(t, result.CreateTeam.JoinCode)
		assert.Equal(t, projectID, result.CreateTeam.ParentProject.ID)
	})

	t.Run("user cannot create team", func(t *testing.T) {
		resp := client.WithAuth(userToken).MustExecute(t, `
			mutation CreateTeam($projectId: ID!, $input: CreateTeamInput!) {
				createTeam(projectId: $projectId, input: $input) {
					id
				}
			}
		`, map[string]any{
			"projectId": projectID,
			"input": map[string]any{
				"name":        "Should Fail",
				"description": "This should not be created",
			},
		})

		require.True(t, resp.HasErrors())
		assert.Contains(t, resp.ErrorMessage(), "unauthorized")
	})

	t.Run("admin can query superteams", func(t *testing.T) {
		resp := client.WithAuth(adminToken).MustExecute(t, `
			query GetSuperTeams($projectId: ID!) {
				superteams(filter: { projectId: $projectId }, first: 100) {
					edges {
						node {
							id
							name
							teams { id name }
						}
					}
					totalCount
				}
			}
		`, map[string]any{"projectId": projectID})

		require.False(t, resp.HasErrors(), "unexpected error: %s", resp.ErrorMessage())

		var result struct {
			Superteams struct {
				Edges []struct {
					Node struct {
						ID    string `json:"id"`
						Name  string `json:"name"`
						Teams []struct {
							ID   string `json:"id"`
							Name string `json:"name"`
						} `json:"teams"`
					} `json:"node"`
				} `json:"edges"`
				TotalCount int `json:"totalCount"`
			} `json:"superteams"`
		}
		require.NoError(t, resp.UnmarshalData(&result))

		// Should have seeded superteams
		assert.Greater(t, result.Superteams.TotalCount, 0)
	})

	t.Run("admin can add team members", func(t *testing.T) {
		// Create a fresh team for this test
		createResp := client.WithAuth(adminToken).MustExecute(t, `
			mutation CreateTeam($projectId: ID!, $input: CreateTeamInput!) {
				createTeam(projectId: $projectId, input: $input) {
					id
				}
			}
		`, map[string]any{
			"projectId": projectID,
			"input": map[string]any{
				"name":        "Member Test Team",
				"description": "Team for testing member additions",
			},
		})
		require.False(t, createResp.HasErrors(), "unexpected error creating team: %s", createResp.ErrorMessage())

		var createResult struct {
			CreateTeam struct{ ID string } `json:"createTeam"`
		}
		require.NoError(t, createResp.UnmarshalData(&createResult))
		newTeamID := createResult.CreateTeam.ID

		// Add members
		userToAdd := data.UserIDs[2] // Use a different user than the admin
		resp := client.WithAuth(adminToken).MustExecute(t, `
			mutation AddMembers($teamId: ID!, $userIds: [ID!]!) {
				addTeamMembers(teamId: $teamId, userIds: $userIds, force: true) {
					id
					members {
						user { id }
					}
				}
			}
		`, map[string]any{
			"teamId":  newTeamID,
			"userIds": []string{userToAdd},
		})

		require.False(t, resp.HasErrors(), "unexpected error: %s", resp.ErrorMessage())

		var result struct {
			AddTeamMembers struct {
				ID      string `json:"id"`
				Members []struct {
					User struct {
						ID string `json:"id"`
					} `json:"user"`
				} `json:"members"`
			} `json:"addTeamMembers"`
		}
		require.NoError(t, resp.UnmarshalData(&result))

		// Verify the user was added
		found := false
		for _, m := range result.AddTeamMembers.Members {
			if m.User.ID == userToAdd {
				found = true
				break
			}
		}
		assert.True(t, found, "user should be in team members")
	})
}

File: elixir_backend/test/elixir_backend_web/schema/teams_test.exs
defmodule ElixirBackendWeb.Schema.TeamsTest do
  use ElixirBackendWeb.ConnCase

  alias ElixirBackend.Teams

  # ── Query: team(id) ──

  describe "team query" do
    test "returns team with members", %{conn: conn} do
      project = create_project()
      {:ok, team} = Teams.create_team(project.id, %{name: "Team Alpha", description: "Desc"})
      user = create_user(%{name: "Member"})
      Teams.add_members(team.id, [user.id])

      query = """
      query($id: ID!) {
        team(id: $id) {
          id name description joinCode leaderboardExcluded
          members { id name isTeamLead joinedAt }
          parentProject { id }
        }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{"id" => team.id})

      data = json_response(conn, 200)["data"]["team"]
      assert data["id"] == team.id
      assert data["name"] == "Team Alpha"
      assert data["joinCode"] != nil
      assert data["leaderboardExcluded"] == false
      assert length(data["members"]) == 1
      assert hd(data["members"])["name"] == "Member"
      assert data["parentProject"]["id"] == project.id
    end

    test "returns error for nonexistent team", %{conn: conn} do
      query = "query($id: ID!) { team(id: $id) { id } }"
      conn = graphql_query(conn, query, %{"id" => "TM00000000000000000000000000"})
      assert json_response(conn, 200)["errors"] != nil
    end
  end

  # ── Query: teamByJoinCode ──

  describe "teamByJoinCode query" do
    test "returns team by join code", %{conn: conn} do
      project = create_project()
      {:ok, team} = Teams.create_team(project.id, %{name: "Team"})
      user = create_user()

      query = """
      query($code: String!, $projectId: ID!) {
        teamByJoinCode(code: $code, projectId: $projectId) { id name }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{"code" => team.join_code, "projectId" => project.id})

      data = json_response(conn, 200)["data"]["teamByJoinCode"]
      assert data["id"] == team.id
    end

    test "returns null for non-existent code", %{conn: conn} do
      project = create_project()
      user = create_user()

      query = """
      query($code: String!, $projectId: ID!) {
        teamByJoinCode(code: $code, projectId: $projectId) { id }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{"code" => "BADCODE1", "projectId" => project.id})

      data = json_response(conn, 200)["data"]["teamByJoinCode"]
      assert data == nil
    end

    test "returns null when code exists but wrong project", %{conn: conn} do
      p1 = create_project(%{name: "P1"})
      p2 = create_project(%{name: "P2"})
      {:ok, team} = Teams.create_team(p1.id, %{name: "Team"})
      user = create_user()

      query = """
      query($code: String!, $projectId: ID!) {
        teamByJoinCode(code: $code, projectId: $projectId) { id }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{"code" => team.join_code, "projectId" => p2.id})

      data = json_response(conn, 200)["data"]["teamByJoinCode"]
      assert data == nil
    end
  end

  # ── Query: teams(filter, pagination) ──

  describe "teams query" do
    test "returns paginated teams filtered by project", %{conn: conn} do
      project = create_project()
      Teams.create_team(project.id, %{name: "T1"})
      Teams.create_team(project.id, %{name: "T2"})
      user = create_user()

      query = """
      query($filter: TeamFilter, $first: Int) {
        teams(filter: $filter, first: $first) {
          edges { node { id name joinCode } }
          totalCount
        }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{
          "filter" => %{"projectId" => project.id},
          "first" => 10
        })

      data = json_response(conn, 200)["data"]["teams"]
      assert data["totalCount"] == 2
      assert length(data["edges"]) == 2
    end
  end

  # ── Mutation: createTeam ──

  describe "createTeam mutation" do
    test "admin can create team", %{conn: conn} do
      project = create_project()
      user = create_user()

      query = """
      mutation($projectId: ID!, $input: CreateTeamInput!) {
        createTeam(projectId: $projectId, input: $input) {
          id name description joinCode
        }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{
          "projectId" => project.id,
          "input" => %{"name" => "New Team", "description" => "A new team"}
        })

      data = json_response(conn, 200)["data"]["createTeam"]
      assert data["name"] == "New Team"
      assert String.starts_with?(data["id"], "TM")
      assert data["joinCode"] != nil
    end

    test "user cannot create team", %{conn: conn} do
      project = create_project()
      user = create_user()

      query = """
      mutation($projectId: ID!, $input: CreateTeamInput!) {
        createTeam(projectId: $projectId, input: $input) { id }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["user"])
        |> graphql_query(query, %{
          "projectId" => project.id,
          "input" => %{"name" => "Unauthorized"}
        })

      assert json_response(conn, 200)["errors"] != nil
    end
  end

  # ── Mutation: addTeamMembers ──

  describe "addTeamMembers mutation" do
    test "admin can add members", %{conn: conn} do
      project = create_project()
      {:ok, team} = Teams.create_team(project.id, %{name: "Team"})
      admin = create_user(%{name: "Admin"})
      member = create_user(%{name: "Member"})

      query = """
      mutation($teamId: ID!, $userIds: [ID!]!) {
        addTeamMembers(teamId: $teamId, userIds: $userIds, force: true) {
          id members { id name }
        }
      }
      """

      conn =
        conn
        |> auth_conn(admin.id, ["admin"])
        |> graphql_query(query, %{"teamId" => team.id, "userIds" => [member.id]})

      data = json_response(conn, 200)["data"]["addTeamMembers"]
      assert length(data["members"]) == 1
      assert hd(data["members"])["name"] == "Member"
    end
  end

  # ── Mutation: joinTeam ──

  describe "joinTeam mutation" do
    test "user can join team via code", %{conn: conn} do
      project = create_project()
      {:ok, team} = Teams.create_team(project.id, %{name: "Team"})
      user = create_user()

      query = """
      mutation($code: ID!) {
        joinTeam(code: $code) { id name }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["user"])
        |> graphql_query(query, %{"code" => team.join_code})

      data = json_response(conn, 200)["data"]["joinTeam"]
      assert data["id"] == team.id
    end

    test "unauthenticated join returns error", %{conn: conn} do
      project = create_project()
      {:ok, team} = Teams.create_team(project.id, %{name: "Team"})

      query = "mutation($code: ID!) { joinTeam(code: $code) { id } }"
      conn = graphql_query(conn, query, %{"code" => team.join_code})
      assert json_response(conn, 200)["errors"] != nil
    end
  end

  # ── Mutation: deleteTeam ──

  describe "deleteTeam mutation" do
    test "admin can delete team", %{conn: conn} do
      project = create_project()
      {:ok, team} = Teams.create_team(project.id, %{name: "Team"})
      user = create_user()

      query = "mutation($id: ID!) { deleteTeam(id: $id) }"

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{"id" => team.id})

      assert json_response(conn, 200)["data"]["deleteTeam"] == true
    end
  end

  # ── SuperTeam queries ──

  describe "superteam query" do
    test "returns super team with teams", %{conn: conn} do
      project = create_project()
      {:ok, st} = Teams.create_super_team(project.id, %{name: "Super Team", description: "Desc"})
      {:ok, team} = Teams.create_team(project.id, %{name: "T1"})
      Teams.assign_teams_to_super_team(st.id, [team.id])

      query = """
      query($id: ID!) {
        superteam(id: $id) {
          id name description
          teams { id name }
          parentProject { id }
        }
      }
      """

      conn = graphql_query(conn, query, %{"id" => st.id})
      data = json_response(conn, 200)["data"]["superteam"]

      assert data["id"] == st.id
      assert data["name"] == "Super Team"
      assert length(data["teams"]) == 1
      assert data["parentProject"]["id"] == project.id
    end
  end

  describe "superteams query" do
    test "returns paginated super teams filtered by project", %{conn: conn} do
      project = create_project()
      Teams.create_super_team(project.id, %{name: "ST1"})
      Teams.create_super_team(project.id, %{name: "ST2"})

      query = """
      query($filter: SuperTeamFilter, $first: Int) {
        superteams(filter: $filter, first: $first) {
          edges { node { id name } }
          totalCount
        }
      }
      """

      conn =
        graphql_query(conn, query, %{
          "filter" => %{"projectId" => project.id},
          "first" => 10
        })

      data = json_response(conn, 200)["data"]["superteams"]
      assert data["totalCount"] == 2
    end
  end

  # ── SuperTeam mutations ──

  describe "createSuperTeam mutation" do
    test "admin can create super team", %{conn: conn} do
      project = create_project()
      user = create_user()

      query = """
      mutation($projectId: ID!, $input: CreateSuperTeamInput!) {
        createSuperTeam(projectId: $projectId, input: $input) {
          id name description color
        }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{
          "projectId" => project.id,
          "input" => %{"name" => "New ST", "description" => "Desc", "color" => "#FF0000"}
        })

      data = json_response(conn, 200)["data"]["createSuperTeam"]
      assert data["name"] == "New ST"
      assert String.starts_with?(data["id"], "ST")
      assert data["color"] == "#FF0000"
    end
  end

  describe "assignTeamsToSuperTeam mutation" do
    test "admin can assign teams to super team", %{conn: conn} do
      project = create_project()
      {:ok, st} = Teams.create_super_team(project.id, %{name: "ST"})
      {:ok, t1} = Teams.create_team(project.id, %{name: "T1"})
      {:ok, t2} = Teams.create_team(project.id, %{name: "T2"})
      user = create_user()

      query = """
      mutation($superTeamId: ID!, $teamIds: [ID!]!) {
        assignTeamsToSuperTeam(superTeamId: $superTeamId, teamIds: $teamIds) {
          id teams { id name }
        }
      }
      """

      conn =
        conn
        |> auth_conn(user.id, ["admin"])
        |> graphql_query(query, %{"superTeamId" => st.id, "teamIds" => [t1.id, t2.id]})

      data = json_response(conn, 200)["data"]["assignTeamsToSuperTeam"]
      assert data["id"] == st.id
      assert length(data["teams"]) == 2
    end
  end
end
